Makefile

```
@@ -312,57 +312,77 @@ undeploy-c9s-%-c9s-python-3.9: bin/kubectl
 	$(info # Undeploying notebook from $(NOTEBOOK_DIR) directory...)
 	$(KUBECTL_BIN) delete -k $(NOTEBOOK_DIR)
 
+# Function for testing readiness by pinging the /api endpoint
+#   ARG 1: Image tag name.
+define test_notebook
+	$(eval FULL_NOTEBOOK_NAME := $(shell $(KUBECTL_BIN) get pods -l app=$(1) -o custom-columns=":metadata.name" | tr -d '\n'))
+	$(info # Running tests for $(1) notebook...)
+	$(KUBECTL_BIN) wait --for=condition=ready pod -l app=$(1) --timeout=600s
+	$(KUBECTL_BIN) port-forward svc/$(1)-notebook 8888:8888 & curl --retry 5 --retry-delay 5 --retry-connrefused http://localhost:8888/notebook/opendatahub/jovyan/api ; EXIT_CODE=$$?; echo && pkill --full "^$(KUBECTL_BIN).*port-forward.*"
+endef
+
+# Function for testing a notebook with papermill  
+#   ARG 1: Notebook name
+#   ARG 1: UBI flavor
+#   ARG 1: Python kernel
+define test_with_papermill
+	$(KUBECTL_BIN) exec $(FULL_NOTEBOOK_NAME) -- /bin/sh -c "wget ${NOTEBOOK_REPO_BRANCH_BASE}/jupyter/$(1)/$(2)-$(3)/test/test_notebook.ipynb -O test_notebook.ipynb && python3 -m papermill test_notebook.ipynb $(1)_$(2)_output.ipynb --kernel python3 --stderr-file $(1)_$(2)_error.txt && cat $(1)_$(2)_error.txt | grep FAILED > /dev/null" ; \
```

With this construct, if the first part of the command fails (`wget` or `papermill` execution), then the second one (`cat`) isn't executed and thus grep doesn't have anything to grep eventually which results into false-negative again. Wouldn't it be better to have it on a separate lines? Something like this:

```suggestion
	$(KUBECTL_BIN) exec $(FULL_NOTEBOOK_NAME) -- /bin/sh -c "wget ${NOTEBOOK_REPO_BRANCH_BASE}/jupyter/$(1)/$(2)-$(3)/test/test_notebook.ipynb -O test_notebook.ipynb && python3 -m papermill test_notebook.ipynb $(1)_$(2)_output.ipynb --kernel python3 --stderr-file $(1)_$(2)_error.txt" ; \
        if [ $$? -ne 0 ]; then \
echo "ERROR: The $(1) $(2) notebook encountered a failure. To investigate the issue, you can review the logs located in the ocp-ci cluster on 'artifacts/notebooks-e2e-tests/jupyter-$(1)-$(2)-$(3)-test-e2e' directory or run 'cat $(1)_$(2)_error.txt' within your container. The make process has been aborted." ; \
		exit 1 ; \
	fi ; \
	
	$(KUBECTL_BIN) exec $(FULL_NOTEBOOK_NAME) -- /bin/sh -c "cat $(1)_$(2)_error.txt | grep --quiet FAILED" ; \
```

---

Good point, I changed it, and it works as expected